### DISCUSSION TITLE
OIDC with Duende IdentityServer

### DISCUSSION THREAD

**Discussion by mweihberg** [datalust/seq-tickets#2028] [Help] [answered]
Hi!
Ive trying to implement oidc authentication (https://docs.datalust.co/docs/openid-connect).
We are using a custom implementation of Duende Identity Server.
Ive managed to register the client and connecting successfully til it reached the user email matching implementation.
It says:
`{"@t":"2023-11-27T16:50:15.4854196Z","@mt":"Identity {UniqueIdentifier} successfully authenticated, but the identity does not carry the email (or equivalent) claim.","@l":"Warning","UniqueIdentifier":"5fabf31a-fab1-43c8-ba4d-100a44fcba0e","Principal":{"Id":"principal-public","OnBehalfOfUserId":null},"ActionId":"63ee685b-5267-4ea4-a6b1-5bf3f8771aa8","ActionName":"Seq.Server.Web.UI.OpenIdConnectController.ChallengeGet (Seq)","RequestId":"0HMVFD3FT30SJ:0000000C","RequestPath":"/oidc/challenge","ConnectionId":"0HMVFD3FT30SJ"}
`
Ive checked the call with postman and received an access token with the desired email claim.
What i am doing wrong? Any help appreciated...

**Comment by mweihberg** [comment, marked as answer]
Solved at my own. :satisfied:

After hours of researching, reading my own question here guided me to the solution.
i was talking about the access token, but important is the **id_token**.

By default Duende IdentityServer not sending user claims along with the id_token. To enabling it you have to enable `AlwaysIncludeUserClaimsInIdToken` at the Client Token Model for the registered seq client. [Reference](https://docs.duendesoftware.com/identityserver/v6/reference/models/client/#token)

**Reply by nblumhardt** [reply]
Fantastic, thanks for sharing your solution, @mweihberg :+1:

**Reply by mweihberg** [reply]
@nblumhardt 
But one thing is a bit strange. The documentation of Duende says about `AlwaysIncludeUserClaimsInIdToken`:

> When requesting both an id token and access token, should the user claims always be added to the id token instead of requiring the client to use the userinfo endpoint. Default is false.

In the logs of the identityserver i can see Seq tries to call userinfo endpoint. But for some reason the response doesnt suffices the needs. Is something wrong with the implementation of [1154](https://github.com/datalust/seq-tickets/issues/1154)?

**Reply by nblumhardt** [reply]
Thanks for the extra info. Is it possible to tell whether the claims coming back from the userinfo endpoint include the email claim? Nothing rings a bell, but if there's anything we can improve it'd be great to know.

**Reply by mweihberg** [reply]
The default response from /connect/userinfo is
```
{
    "sub": "5fabf31a-fab1-43c8-ba4d-100a44fcba0e",
    "preferred_username": "",
    "unique_name": "",
    "last_name": "",
    "fore_name": "",
    "name": "",
    "email": "",
    "email_verified": true
}
```
Obtained with postman via the authorization_code flow and the same clientid/secret as the registered seq client.

**Reply by nblumhardt** [reply]
Thanks @mweihberg - looks like `email` is empty there, so may be an IdentityServer configuration issue.

**Reply by mweihberg** [reply]
No, sorry i havent told you, that due to privacy reasons ive emptied the response.
I thought we are talking about the general structure of the response model...

**Reply by nblumhardt** [reply]
Thanks for the extra info; I'll keep an eye on it :+1:
